export type ApplyKnockback = typeof(require("@game/ReplicatedFirst/renouveau/packages/applyKnockback"))
export type Controls = typeof(require("@game/ReplicatedFirst/renouveau/packages/controls"))
export type CoreGui = typeof(require("@game/ReplicatedFirst/renouveau/packages/coreGui"))
export type FormatTime = typeof(require("@game/ReplicatedFirst/renouveau/packages/formatTime"))
export type SoundUtils = typeof(require("@game/ReplicatedFirst/renouveau/packages/soundUtils"))
export type StateMachine = typeof(require("@game/ReplicatedFirst/renouveau/packages/stateMachine"))
export type ToRGB = typeof(require("@game/ReplicatedFirst/renouveau/packages/toRGB"))
export type TweenUtils = typeof(require("@game/ReplicatedFirst/renouveau/packages/tweenUtils"))
export type Weld = typeof(require("@game/ReplicatedFirst/renouveau/packages/weld"))
return {}